feat: Implement barebones Forum channel support.

FayeDel opened the pull request:
## About

This PR title essentially explains itself.
Forums are in testing but has been more widespread, so this is left as a draft PR until its live/more features are discovered.

This correlates to an earlier PR: #630 

## Checklist

- [X] I've ran `pre-commit` to format and lint the change(s) made.
- [X] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [X] I've made this pull request for/as: (check all that apply)
  - [X] Documentation
  - [X] Breaking change
  - [X] New feature/enhancement
  - [ ] Bugfix

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
Merge branch 'interactions-py:unstable' into unstable

FayeDel committed:
docs, feat: Implement barebones Forum channel support.

FayeDel committed:
docs: Update channel attributes.

FayeDel committed:
feat: Implement barebones create thread in forum function.

FayeDel committed:
feat: Implement tags support, implement creating post in Forums.

FayeDel committed:
fix: Include headers.

FayeDel committed:
chore: Rebase from unstable, convert to attrs

FayeDel committed:
chore: Remove redundant slots, update flag headers.

FayeDel committed:
chore: Update PR to latest unstable.

i0bs commented:
As of 5 July.

![image](https://user-images.githubusercontent.com/41456914/178174621-7850b17e-445b-48ed-9152-2913de85cb5e.png)

FayeDel committed:
chore: Update PR to latest unstable.

FayeDel committed:
Merge branch 'unstable--upstream' into unstable-undoc-forums

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -1201,10 +1228,60 @@ async def join(self) -> None:
+    ...
+
+
+@define()
+class Forum(Thread):
+    """An object representing a Forum.
+    .. note::
+        While this isn't directly dispatched for GW based events,
+        this object exists moreso for future helper methods.
```
Wouldn't it make more sense to keep the methods on the channel object like we did for threads?

FayeDel committed:
chore: Update to latest unstable commit.

(This may fail ci because of merge conflicts with documentation using "=======")

FayeDel committed:
feat: Apply previous forum breaking change, document tags in http method.

FayeDel commented:
Forums are officially out so I'm marking this as ready for review ^^

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -95,6 +96,22 @@ class ThreadMember(ClientSerializerMixin):
     :ivar bool muted: Whether the member is muted or not.
     """
 
+    # __slots__ = (
+    #     "_json",
+    #     "id",
+    #     "user_id",
+    #     "join_timestamp",
+    #     "flags",
+    #     # TODO: Document below attributes.
+    #     "muted",  # ??
+    #     "user",
+    #     "team_id",
+    #     "membership_state",
+    #     "permissions",
+    #     "member",
+    #     "presence",
+    # )
```
_uh_

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -104,6 +121,12 @@ class ThreadMember(ClientSerializerMixin):
 
+    # Take these conversions with a grain of salt.
+
+    # Commented out for circular import reasons, but schema says they do work.
+    # self.member = Member(**self._json.get("member")) if self._json.get("member") else None
+    # self.presence = Presence(**self._json.get("presence")) if self._json.get("presence") else None
```
Will you do something with this or just let it exist?

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -319,6 +342,15 @@ class Channel(ClientSerializerMixin, IDMixin):
     :ivar Optional[int] total_message_sent?: Number of messages ever sent in a thread.
     """
 
+    # __slots__ = (
+    #     # TODO: Document banner when Discord officially documents them.
+    #     "banner",
+    #     "guild_hashes",
+    #     "flags",
+    #     "template",
+    #     "available_tags",
+    # )
```
_uh_

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -1505,10 +1537,60 @@ async def get_permissions_for(self, member: "Member") -> Permissions:
+
+    # __slots__ = (
+    #     "_json",
+    #     "id",
+    #     "type",
+    #     "guild_id",
+    #     "position",
+    #     "permission_overwrites",
+    #     "name",
+    #     "topic",
+    #     "nsfw",
+    #     "last_message_id",
+    #     "bitrate",
+    #     "user_limit",
+    #     "rate_limit_per_user",
+    #     "recipients",
+    #     "icon",
+    #     "owner_id",
+    #     "application_id",
+    #     "parent_id",
+    #     "last_pin_timestamp",
+    #     "rtc_region",
+    #     "video_quality_mode",
+    #     "message_count",
+    #     "member_count",
+    #     "thread_metadata",
+    #     "member",
+    #     "default_auto_archive_duration",
+    #     "permissions",
+    #     "_client",
+    #     # TODO: Document these when Discord officially documents them.
+    #     "banner",
+    #     "guild_hashes",
+    #     "flags",  # This seems to be forum unique.
+    #     "newly_created",
+    #     "template",
+    #     "available_tags",
+    # )
+
+    ...
```
_uh_

sillyElvy commented on interactions/api/models/channel.py:
```diff
@@ -1505,10 +1537,60 @@ async def get_permissions_for(self, member: "Member") -> Permissions:
+
+@define()
+class Forum(Thread):
+    """An object representing a Forum.
+    .. note::
+        While this isn't directly dispatched for GW based events,
+        this object exists moreso for future helper methods.
```
Helpers are generally implemented in the base channel class so people don't have to use different objects for different channels - if this is not needed, it should be removed

sillyElvy commented on interactions/api/models/gw.py:
```diff
@@ -517,6 +517,10 @@ class MessageReactionRemove(MessageReaction):
     # todo see if the missing member attribute affects anything
 
 
+# TODO: (temp) Add Thread object for a THREAD_UPDATE event.
+# More prevalent on forums for some odd reason.
```
If its forum related why not add it?

sillyElvy commented:
Docs build failed

FayeDel committed:
feat: Implement forum tags object, document forum-specific attributes in Channel model, fix MISSING import in thread.py

FayeDel committed:
fix: Fix tags attrs init typo.

FayeDel committed:
chore: Remove uncommented __slots__ in ThreadMember

The slots code, for the most part, were from when forums were in early beta (when thread related updates were dispatched quickly). Now the attributes doesn't seem to exist anymore.

sillyElvy reviewed (approved):
Lgtm!
